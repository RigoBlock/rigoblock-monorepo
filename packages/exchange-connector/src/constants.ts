export const MAINNET: string = 'mainnet'
export const ROPSTEN: string = 'ropsten'
export const KOVAN: string = 'kovan'
export const RINKEBY: string = 'rinkeby'

export const PRICE_PRECISION: number = 7
export const AMOUNT_PRECISION: number = 5
export const TO_WEI: number = 1e18

export enum supportedExchanges {
  ETHFINEX = 'Ethfinex',
  ETHFINEX_RAW = 'EthfinexRaw',
  ERCDEX = 'ERCdEX',
  ZEROEXRELAYER = '0xStandardRelayer'
}

export enum NETWORKS {
  MAINNET = 1,
  KOVAN = 42,
  ROPSTEN = 3,
  RINKEBY = 4
}

export enum WS_STATUS {
  CLOSED = 'closed',
  OPEN = 'open',
  CONNECTING = 'connecting',
  CLOSING = 'closing'
}
