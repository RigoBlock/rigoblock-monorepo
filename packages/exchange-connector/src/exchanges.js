// Copyright 2017 Rigo Investment Sagl.
// This file is part of RigoBlock.

import { NETWORKS_ID, SupportedExchanges } from './const'

export const getTickers = {
  ERCdEX: (networkId = 1) => {
    const options = {
      method: 'GET',
      url: `${
        SupportedExchanges.ERCdEX.http[NETWORKS_ID[networkId]]
      }/reports/ticker?networkId=${networkId}`,
      qs: {},
      json: true
    }
    return options
  },
  Ethfinex: (networkId = 1) => {
    const symbols = SupportedExchanges.Ethfinex.tickersTokenPairs.toString()
    const options = {
      method: 'GET',
      url: `${
        SupportedExchanges.Ethfinex.http[NETWORKS_ID[networkId]]
      }/tickers?symbols=${symbols}`,
      qs: {},
      json: true
    }
    return options
  }
}

export const getAggregatedOrders = {
  ERCdEX: (networkId = 1, baseTokenAddress, quoteTokenAddress) => {
    const options = {
      method: 'GET',
      uri: `${
        SupportedExchanges.ERCdEX.http[NETWORKS_ID[networkId]]
      }/aggregated_orders`,
      qs: {
        networkId: networkId,
        baseTokenAddress: baseTokenAddress,
        quoteTokenAddress: quoteTokenAddress
      },
      json: true
    }
    return options
  },
  Ethfinex: (networkId = 1, baseToken, quoteToken) => {
    if (baseToken.length > 4) {
      throw new Error('baseToken has to be a shorter than 4 characters')
    }
    if (quoteToken.length > 4) {
      throw new Error('quoteToken has to be a shorter than 4 characters')
    }
    const options = {
      method: 'GET',
      url: `${
        SupportedExchanges.Ethfinex.http[NETWORKS_ID[networkId]]
      }/book/t${baseToken}${quoteToken}/P0`,
      qs: {},
      json: true
    }
    return options
  }
}
