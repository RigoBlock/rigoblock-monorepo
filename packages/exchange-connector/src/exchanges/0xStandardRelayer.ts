export class StandardRelayer {
  constructor(public API_URL, public transport = 'http') {}

  /**
   * Accepts one or two optional token addresses.
   * Setting only tokenA or tokenB returns pairs filtered by that token only
   */
  public async getTokenPairs(
    tokenA?: string,
    tokenB?: string
  ): Promise<StandardRelayer.rawTokenPairs> {
    const url = `${
      this.API_URL
    }/v0/token_pairs?tokenA=${tokenA}&tokenB=${tokenB}`
    return fetch(url).then(r => r.json())
  }

  public async getOrders(
    exchangeContractAddress?: string, // returns orders created for this exchange address
    tokenAddress?: string, // returns orders where makerTokenAddress or takerTokenAddress is token address,
    makerTokenAddress?: string, // returns orders with specified makerTokenAddress,
    takerTokenAddress?: string, // returns orders with specified makerTokenAddress,
    maker?: string, // returns orders where maker is maker address,
    taker?: string, // returns orders where taker is taker address,
    trader?: string, // returns orders where maker or taker is trader address,
    feeRecipient?: string // returns orders where feeRecipient is feeRecipient address
  ): Promise<StandardRelayer.rawOrder[]> {
    const url =
      `${
        this.API_URL
      }/v0/orders?exchangeContractAddress=${exchangeContractAddress}&tokenAddress=${tokenAddress}` +
      `&makerTokenAddress=${makerTokenAddress}&takerTokenAddress=${takerTokenAddress}&maker=${maker}` +
      `&taker=${taker}&trader=${trader}&feeRecipient=${feeRecipient}`
    return fetch(url).then(r => r.json())
  }

  public async getOrder(orderHash: string): Promise<StandardRelayer.rawOrder> {
    const url = `${this.API_URL}/v0/order/${orderHash}`
    return fetch(url).then(r => r.json())
  }

  public async getOrderbook(
    baseTokenAddress: string, // address of token designated as the baseToken in the currency pair calculation of price
    quoteTokenAddress: string // address of token designated as the quoteToken in the currency pair calculation of price
  ): Promise<StandardRelayer.orderBook> {
    const url = `${
      this.API_URL
    }/v0/orderbook?baseTokenAddress=${baseTokenAddress}&quoteTokenAddress=${quoteTokenAddress}`
    return fetch(url).then(r => r.json())
  }
}

export namespace StandardRelayer {
  export type tokenPair = {
    tokenA: {
      address: string
      minAmount: string
      maxAmount: string
      precision: number
    }
    tokenB: {
      address: string
      minAmount: string
      maxAmount: string
      precision: number
    }
  }

  export type rawOrder = {
    exchangeContractAddress: string
    maker: string
    taker: string
    makerTokenAddress: string
    takerTokenAddress: string
    feeRecipient: string
    makerTokenAmount: string
    takerTokenAmount: string
    makerFee: string
    takerFee: string
    expirationUnixTimestampSec: string
    salt: string
    ecSignature: {
      v: number
      r: string
      s: string
    }
    remainingTakerTokenAmount: string
    orderHash: string
    source: string
    price: string
  }

  export type rawTokenPairs = tokenPair[]

  export type orderBook = {
    bids: rawOrder[]
    asks: rawOrder[]
  }
}

export default StandardRelayer
