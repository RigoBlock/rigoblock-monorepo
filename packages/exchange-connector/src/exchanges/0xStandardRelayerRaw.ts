export class StandardRelayerRaw<T = StandardRelayer.RawOrder[]> {
  constructor(public STANDARD_API_URL, public transport = 'http') {
    if (!STANDARD_API_URL) {
      throw new Error('API url must be specified.')
    }
  }

  /**
   * Accepts one or two optional token addresses.
   * Setting only tokenA or tokenB returns pairs filtered by that token only
   */
  public async getTokenPairs(
    tokenA?: string,
    tokenB?: string
  ): Promise<StandardRelayer.TokenPair[]> {
    const url = `${
      this.STANDARD_API_URL
    }/v0/token_pairs?tokenA=${tokenA}&tokenB=${tokenB}`
    return fetch(url).then(r => r.json())
  }

  public async getOrders(
    exchangeContractAddress?: string, // returns orders created for this exchange address
    tokenAddress?: string, // returns orders where makerTokenAddress or takerTokenAddress is token address,
    makerTokenAddress?: string, // returns orders with specified makerTokenAddress,
    takerTokenAddress?: string, // returns orders with specified makerTokenAddress,
    maker?: string, // returns orders where maker is maker address,
    taker?: string, // returns orders where taker is taker address,
    trader?: string, // returns orders where maker or taker is trader address,
    feeRecipient?: string // returns orders where feeRecipient is feeRecipient address
  ): Promise<StandardRelayer.RawOrder[]> {
    const url = `${
      this.STANDARD_API_URL
    }/v0/orders?exchangeContractAddress=${exchangeContractAddress}&tokenAddress=${tokenAddress}
        &makerTokenAddress=${makerTokenAddress}&takerTokenAddress=${takerTokenAddress}&maker=${maker}
        &taker=${taker}&trader=${trader}&feeRecipient=${feeRecipient}`
    return fetch(url).then(r => r.json())
  }

  public async getOrder(orderHash: string): Promise<StandardRelayer.RawOrder> {
    const url = `${this.STANDARD_API_URL}/v0/order/${orderHash}`
    return fetch(url).then(r => r.json())
  }

  public async getOrderbook(
    baseTokenAddress: string, // address of token designated as the baseToken in the currency pair calculation of price
    quoteTokenAddress: string // address of token designated as the quoteToken in the currency pair calculation of price
  ): Promise<StandardRelayer.OrderBook> {
    const url = `${
      this.STANDARD_API_URL
    }/v0/orderbook?baseTokenAddress=${baseTokenAddress}&quoteTokenAddress=${quoteTokenAddress}`
    return fetch(url).then(r => r.json())
  }
}

export namespace StandardRelayer {
  export type TokenPair = {
    tokenA: {
      address: string
      minAmount: string
      maxAmount: string
      precision: number
    }
    tokenB: {
      address: string
      minAmount: string
      maxAmount: string
      precision: number
    }
  }

  export type RawOrder = {
    exchangeContractAddress: string
    maker: string
    taker: string
    makerTokenAddress: string
    takerTokenAddress: string
    feeRecipient: string
    makerTokenAmount: string
    takerTokenAmount: string
    makerFee: string
    takerFee: string
    expirationUnixTimestampSec: string
    salt: string
    ecSignature: {
      v: number
      r: string
      s: string
    }
    remainingTakerTokenAmount: string
    orderHash: string
    source: string
    price: string
  }

  export type OrderBook = {
    bids: RawOrder[]
    asks: RawOrder[]
  }
}

export default StandardRelayer
