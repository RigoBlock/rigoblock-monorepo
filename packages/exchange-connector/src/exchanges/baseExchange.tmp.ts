import * as ExchangeTypes from './types'
import { NETWORKS, SupportedExchanges } from '../constants'

export default abstract class BaseExchange<T> {
  public switch(
    exchangeName: SupportedExchanges,
    networkId: NETWORKS = NETWORKS.MAINNET,
    transport: string = 'http',
    exchangesMap: ExchangeTypes.ExchangesMap<T>
  ): T {
    const selectedExchange = exchangesMap[exchangeName]
    if (!selectedExchange) {
      throw new Error(`Exchange ${exchangeName} is not supported.`)
    }
    if (!selectedExchange.supportedNetworks.includes(networkId)) {
      throw new Error(`Network not supported on this exchange: ${networkId}`)
    }

    return new selectedExchange(networkId, transport)
  }
}
