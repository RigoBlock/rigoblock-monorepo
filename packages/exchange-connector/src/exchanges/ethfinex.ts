import { AMOUNT_PRECISION, NETWORKS, PRICE_PRECISION } from '../constants'
import { OrderType, OrdersList, TickersList } from './types'
import BigNumber from 'bignumber.js'
import EthfinexRaw from './ethfinexRaw'

export class Ethfinex {
  static SUPPORTED_NETWORKS: NETWORKS[] = [NETWORKS.MAINNET, NETWORKS.ROPSTEN, NETWORKS.KOVAN]
  private raw: EthfinexRaw
  public options

  constructor(
    public NETWORK_ID: NETWORKS | number,
    public HTTP_URL?: string,
    public WS_URL?: string
  ) {
    this.raw = new EthfinexRaw(NETWORK_ID, HTTP_URL, WS_URL)
    this.options = this.raw.options
  }

  public http = {
    getOrders: async (options: {
      symbols: string
      precision?: EthfinexRaw.OrderPrecisions
    }): Promise<OrdersList> => {
      return this.raw.http
        .getOrders(options)
        .then(result => this.formatOrders(result))
    },
    getTickers: (options: { symbols: string[] }): Promise<TickersList> => {
      return this.raw.http
        .getTickers(options)
        .then(result => this.formatTickers(result))
    },
    getCandles: (options: {
      timeframe: EthfinexRaw.CandlesTimeFrame
      symbols: string
      section: EthfinexRaw.CandlesSection
      limit?: string // max number of candles we want to receive
      sort?: EthfinexRaw.CandlesSort
      start?: string // filter start (ms)
      end?: string // filter end (ms)
    }): Promise<EthfinexRaw.RawCandle[]> => {
      return this.raw.http.getCandles(options)
    }
  }
  public ws = {
    open: () => {
      return this.raw.ws.open()
    },
    close: () => {
      return this.raw.ws.close()
    },
    getConnection: () => {
      return this.raw.ws.getConnection()
    },
    getTickers: async (
      options: { symbols: string[] },
      callback: (err: Error, message?: any) => any
    ) => {
      return this.raw.ws.getTickers(options, callback)
    },
    getCandles: async (
      options: {
        timeframe: string
        symbols: string
      },
      callback: (err: Error, message?: any) => any
    ) => {
      return this.raw.ws.getCandles(options, callback)
    }
  }

  private formatOrders(orders: EthfinexRaw.RawOrder[]): any {
    if (this.checkForError(orders as any)) {
      throw new Error(orders.pop() as any)
    }
    const bids = []
    const asks = []
    let spread
    // format orders and divide between asks and bids
    orders.map((order: number[]) => {
      const type =
        order.slice(order.length - 1).pop() > 0 ? OrderType.BUY : OrderType.SELL
      const orderPrice = new BigNumber(order.shift()).toFixed(PRICE_PRECISION)
      const ordersCount: number = order.shift()
      const orderAmount = new BigNumber(order.pop())
        .absoluteValue()
        .toFixed(AMOUNT_PRECISION)
      const formattedOrder = {
        orderPrice,
        orderAmount,
        ordersCount
      }
      return type === OrderType.BUY
        ? bids.push(formattedOrder)
        : asks.push(formattedOrder)
    })
    // calculate spread
    if (bids.length && asks.length) {
      spread = new BigNumber(asks[0].orderPrice)
        .minus(new BigNumber(bids[0].orderPrice))
        .toFixed(6)
    } else {
      spread = new BigNumber(0).toFixed(6)
    }
    return {
      aggregated: true,
      asks: asks.reverse(),
      bids,
      spread
    }
  }

  private formatTickers(tickers: EthfinexRaw.RawTicker[]): TickersList {
    if (this.checkForError(tickers as any)) {
      throw new Error(tickers.pop() as any)
    }
    if (!tickers.length) {
      throw new Error('Tokens Pair not recognised by exchange')
    }
    const tickersList = tickers.map(ticker => ({
      symbol: ticker[0].substring(1, 4),
      priceEth: ticker[7].toString(),
      priceUsd: ''
    }))
    tickersList.push({
      priceEth: '1',
      priceUsd: '',
      symbol: 'WETH'
    })
    tickersList.push({
      priceEth: '1',
      priceUsd: '',
      symbol: 'ETH'
    })
    return tickersList
  }

  public network(id: number = NETWORKS.MAINNET): Ethfinex {
    return new Ethfinex(id)
  }

  private checkForError(array: any[]) {
    return array[0] === ('error' as any)
  }
}

export default Ethfinex
