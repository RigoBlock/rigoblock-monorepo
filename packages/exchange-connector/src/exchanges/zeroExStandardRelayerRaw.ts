import { NETWORKS } from '../constants'
import { fetchJSON, getQueryParameters, postJSON } from '../utils'

export class ZeroExStandardRelayerRaw<T = ZeroExStandardRelayerRaw.RawOrder[]> {
  static SUPPORTED_NETWORKS: NETWORKS[] = [
    NETWORKS.MAINNET,
    NETWORKS.KOVAN,
    NETWORKS.ROPSTEN
  ]
  constructor(
    public networkId: NETWORKS | number,
    public STANDARD_HTTP_URL: string,
    public STANDARD_WS_URL?: string
  ) {
    if (!STANDARD_HTTP_URL) {
      throw new Error('API url must be specified for 0x standard relayers.')
    }
  }
  public http = {
    /**
     * Accepts one or two optional token addresses.
     * Setting only tokenA or tokenB returns pairs filtered by that token only
     */
    getTokenPairs: async (options: {
      tokenA?: string
      tokenB?: string
    }): Promise<ZeroExStandardRelayerRaw.TokenPair[]> => {
      const url = `${this.STANDARD_HTTP_URL}/v0/token_pairs`
      const queryParams = getQueryParameters(options)
      return fetchJSON(url, queryParams)
    },
    getOrders: async (options: {
      exchangeContractAddress?: string
      tokenAddress?: string
      makerTokenAddress?: string
      takerTokenAddress?: string
      maker?: string
      taker?: string
      trader?: string
      feeRecipient?: string
    }): Promise<ZeroExStandardRelayerRaw.RawOrder[]> => {
      const url = `${this.STANDARD_HTTP_URL}/v0/orders`
      const queryParams = getQueryParameters(options)
      return fetchJSON(url, queryParams)
    },
    getOrder: async (options: {
      orderHash: string
    }): Promise<ZeroExStandardRelayerRaw.RawOrder> => {
      const url = `${this.STANDARD_HTTP_URL}/v0/order/${options.orderHash}`
      return fetchJSON(url)
    },
    getOrderbook: async (options: {
      baseTokenAddress: string
      quoteTokenAddress: string
    }): Promise<ZeroExStandardRelayerRaw.OrderBook> => {
      const url = `${this.STANDARD_HTTP_URL}/v0/orderbook`
      const queryParams = getQueryParameters(options)
      return fetchJSON(url, queryParams)
    },
    getFees: async (options: {
      exchangeContractAddress: string
      maker: string
      taker: string
      makerTokenAddress: string
      takerTokenAddress: string
      makerTokenAmount: string
      takerTokenAmount: string
      expirationUnixTimestampSec: string
      salt: string
    }): Promise<ZeroExStandardRelayerRaw.RawFee> => {
      const url = `${this.STANDARD_HTTP_URL}/v0/fees`
      return postJSON(url, options)
    },
    createOrder: async (options: {
      exchangeContractAddress: string
      maker: string
      taker: string
      makerTokenAddress: string
      takerTokenAddress: string
      feeRecipient: string
      makerTokenAmount: string
      takerTokenAmount: string
      makerFee: string
      takerFee: string
      expirationUnixTimestampSec: string
      salt: string
      ecSignature: {
        v: number
        r: string
        s: string
      }
    }): Promise<ZeroExStandardRelayerRaw.OrderReceipt> => {
      const url = `${this.STANDARD_HTTP_URL}/v0/order`
      return postJSON(url, options)
    }
  }
  public ws = {}
}

export namespace ZeroExStandardRelayerRaw {
  export type TokenPair = {
    tokenA: {
      address: string
      minAmount: string
      maxAmount: string
      precision: number
    }
    tokenB: {
      address: string
      minAmount: string
      maxAmount: string
      precision: number
    }
  }

  export type RawFee = {
    feeRecipient: string
    makerFee: string
    takerFee: string
  }

  export type OrderReceipt = {
    price: string
    exchangeContractAddress: string
    expirationUnixTimestampSec: number
    feeRecipient: string
    maker: string
    makerFee: string
    makerTokenAddress: string
    makerTokenAmount: string
    networkId: number
    orderHash: string
    remainingTakerTokenAmount: string
    salt: string
    serializedEcSignature: string
    state: string
    taker: string
    takerFee: string
    takerTokenAddress: string
    takerTokenAmount: string
    accountId: number
    source: string
    dateUpdated: string
    dateCreated: string
    dateClosed: null
    id: number
  }

  export type RawOrder = {
    exchangeContractAddress: string
    maker: string
    taker: string
    makerTokenAddress: string
    takerTokenAddress: string
    feeRecipient: string
    makerTokenAmount: string
    takerTokenAmount: string
    makerFee: string
    takerFee: string
    expirationUnixTimestampSec: string
    salt: string
    ecSignature: {
      v: number
      r: string
      s: string
    }
    remainingTakerTokenAmount: string
    orderHash: string
    source: string
    price: string
  }

  export type OrderBook = {
    bids: RawOrder[]
    asks: RawOrder[]
  }
}

export default ZeroExStandardRelayerRaw
