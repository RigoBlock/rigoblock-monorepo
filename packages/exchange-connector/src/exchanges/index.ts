import { Ethfinex } from './ethfinex'
import { EthfinexRaw } from './ethfinexRaw'
import { ZeroExStandardRelayerRaw } from './zeroExStandardRelayerRaw'
import { supportedExchanges } from '../constants'

export default {
  [supportedExchanges.ETHFINEX]: Ethfinex,
  [supportedExchanges.ETHFINEX_RAW]: EthfinexRaw,
  [supportedExchanges.ZEROEXRELAYER]: ZeroExStandardRelayerRaw
}
