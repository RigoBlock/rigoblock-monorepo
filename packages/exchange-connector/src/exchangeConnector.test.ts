import { Ethfinex } from './exchanges/ethfinex'
import { NETWORKS, supportedExchanges } from './constants'
import { ZeroExStandardRelayerRaw } from './exchanges/zeroExStandardRelayerRaw'
import ExchangeConnector from './exchangeConnector'

describe('Exchange factory class', () => {
  it('returns a new instance of the selected Exchange class', () => {
    const connector = new ExchangeConnector()
    const exchange = connector.getExchange(supportedExchanges.ETHFINEX)
    expect(exchange).toBeInstanceOf(Ethfinex)
  })

  it('works as a singleton', () => {
    const connector = new ExchangeConnector()
    const exchange = connector.getExchange(supportedExchanges.ETHFINEX)
    const secondExchange = connector.getExchange(supportedExchanges.ETHFINEX)
    expect(secondExchange).toBe(exchange)
  })

  it('throws an error if we try to instance an unsupported exchange', () => {
    const connector = new ExchangeConnector()
    expect(() =>
      connector.getExchange('random' as any)
    ).toThrowErrorMatchingSnapshot()
  })

  it('throws an error if we try to instance an exchange with an unsupported network', () => {
    const connector = new ExchangeConnector()
    expect(() =>
      connector.getExchange(supportedExchanges.ETHFINEX, {
        networkId: NETWORKS.RINKEBY
      })
    ).toThrowErrorMatchingSnapshot()
  })

  it('instantiates the ZeroExStandardRelayerRaw if we pass "0xStandardRelayer" as exchangeName and an api url', () => {
    const connector = new ExchangeConnector()
    const exchange = connector.getExchange(supportedExchanges.ZEROEXRELAYER, {
      httpUrl: 'https://api.radarrelay.com/0x',
      networkId: NETWORKS.MAINNET
    })
    expect(exchange).toBeInstanceOf(ZeroExStandardRelayerRaw)
  })

  it('throws an error if we try to instantiate a 0x relayer without passing the API url', () => {
    const connector = new ExchangeConnector()
    expect(() =>
      connector.getExchange(supportedExchanges.ZEROEXRELAYER as any)
    ).toThrowErrorMatchingSnapshot()
  })
})
