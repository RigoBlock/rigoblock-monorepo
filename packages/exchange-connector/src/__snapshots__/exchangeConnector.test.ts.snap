// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Exchange factory class throws an error if we try to instance an exchange with an unsupported network 1`] = `"Network Id not supported for selected network: 4"`;

exports[`Exchange factory class throws an error if we try to instance an unsupported exchange 1`] = `"Exchange not supported: random"`;

exports[`Exchange factory class throws an error if we try to instantiate a 0x relayer without passing the API url 1`] = `"API url must be specified for 0x standard relayers."`;
